File: Haskell/euler009/test/Euler009Test.hs
import Test.Tasty
import Test.Tasty.HUnit

import Euler009

unitTests :: TestTree
unitTests = testGroup "Euler - 009: Unit Tests"
  [ testCase
      "A 3-4-5 Pythagorean triplet has a sum of 12" $
      findTripletWithSumN 12 @?= Just [3, 4, 5]

  , testCase
      "A 3-4-5 Pythagorean triplet has a sum of 12 and a product of 60" $
      findTripletProductWithSumN 12 @?= Just 60

  , testCase
      "Final Answer (A): 375-200-425 is a Pythagorean triplet for which \
      \a + b + c = 1000" $
      findTripletWithSumN 1000 @?= Just [375, 200, 425]

  , testCase
      "Final Answer (B): The product abc of the Pythagorean triplet for which \
      \a + b + c = 1000 is 31875000" $
      euler009 @?= 31875000
  ]

main :: IO()
main = defaultMain unitTests

File: CXX/tests/euler009_test.cpp
/*
 * Project Euler - 009 - Special Pythagorean triplet
 * Tamara Roberson <tamara.roberson@gmail.com>
 * Copyright (c) 2019 Tamara Roberson
 *
 * A Pythagorean triplet is a set of three natural numbers, a < b < c,
 * for which,
 *
 *      a^2 + b^2 = c^2
 *
 * For example,
 *
 *      3^2 + 4^2 = 9 + 16 = 25 = 5^2.
 *
 * There exists exactly one Pythagorean triplet for which a + b + c = 1000.
 *
 * Find the product abc.
 */

#include <euler/euler009.hpp>

#include <vector> // std::vector

#include <catch2/catch.hpp>


namespace euler
{

// find_pythagorean_triplet_with_sum(sum)
TEST_CASE("A 3-4-5 Pythagorean triplet has a sum of 12")
{
    auto triplet = std::vector<int> { 3, 4, 5 };
    REQUIRE(find_pythagorean_triplet_with_sum(12) == triplet);
}

TEST_CASE("Find a Pythagorean triplet with a sum of 1000")
{
    auto triplet = std::vector<int> { 375, 200, 425 };
    REQUIRE(find_pythagorean_triplet_with_sum(1000) == triplet);
}


// euler009()
TEST_CASE("Euler 009: Final answer is correct")
{
    REQUIRE(euler009() == 31875000);
}

} // end namespace euler
